iree/hal/metal/metal_kernel_library.mm

```diff
@@ -0,0 +1,135 @@
+    id<MTLDevice> device, ExecutableCachingModeBitfield mode,
+    const MetalExecutableDef& metal_executable_def) {
+  IREE_TRACE_SCOPE0("MetalKernelLibrary::Create");
+  if (!metal_executable_def.entry_points() || metal_executable_def.entry_points()->size() == 0) {
+    return InvalidArgumentErrorBuilder(IREE_LOC) << "No entry points defined";
+  }
+  if (!metal_executable_def.shader_sources() ||
+      metal_executable_def.shader_sources()->size() == 0) {
+    return InvalidArgumentErrorBuilder(IREE_LOC) << "No MSL source string present";
+  }
+
+  const auto& entry_points = *metal_executable_def.entry_points();
+  const auto& msl_sources = *metal_executable_def.shader_sources();
+
+  if (entry_points.size() != msl_sources.size()) {
+    return InvalidArgumentErrorBuilder(IREE_LOC)
+           << "Number of entry points and source strings mismatch";
+  }
+
+  // Compile each MSL source string into a MTLLibrary and get the MTLFunction for the entry point to
+  // build the pipeline state object.
+
+  absl::InlinedVector<id<MTLLibrary>, 1> libraries;
+  absl::InlinedVector<id<MTLFunction>, 1> functions;
+  absl::InlinedVector<id<MTLComputePipelineState>, 1> states;
+
+  MTLCompileOptions* msl_compile_options = [MTLCompileOptions new];
+  msl_compile_options.languageVersion = MTLLanguageVersion2_0;
+
+  auto cleanup = MakeCleanup([&]() {
+    for (id<MTLComputePipelineState> state : states) [state release];
+    for (id<MTLFunction> function : functions) [function release];
+    for (id<MTLLibrary> library : libraries) [library release];
+    [msl_compile_options release];
+  });
+
+  for (int i = 0; i < msl_sources.size(); ++i) {
+    @autoreleasepool {
+      NSError* error = nil;
+
+      NSString* shader_source = [NSString stringWithCString:msl_sources[i]->c_str()
+                                                   encoding:[NSString defaultCStringEncoding]];
+      id<MTLLibrary> library = [device newLibraryWithSource:shader_source
+                                                    options:msl_compile_options
+                                                      error:&error];
+      if (!library) {
+        NSLog(@"Failed to create MTLLibrary: %@", error);
+        NSLog(@"Original MSL source: %@", shader_source);
+        return InvalidArgumentErrorBuilder(IREE_LOC) << "Invalid MSL source";
+      }
+      libraries.push_back(library);
+
+      NSString* entry_point = [NSString stringWithCString:entry_points[i]->c_str()
+                                                 encoding:[NSString defaultCStringEncoding]];
+      id<MTLFunction> function = [library newFunctionWithName:entry_point];
+      if (!function) {
+        NSLog(@"Failed to create MTLFunction");
+        NSLog(@"Original MSL source: %@", shader_source);
+        return InvalidArgumentErrorBuilder(IREE_LOC)
+               << "Cannot find entry point '" << entry_points[i] << "' in the " << i
+               << "th shader source";
```

Comment:
nit: this will read strangely for some values of `i` (`0th`, `1th`, `2th`, `3th`).

How about this?
```suggestion
               << "Cannot find entry point '" << entry_points[i] << "' in shader source index "
               << i;
```